src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs

```diff
@@ -85,8 +89,16 @@ private void AddPageModels(PageRouteModelProviderContext context)
                     continue;
                 }
 
-                var routeModel = new PageRouteModel(relativePath, viewEnginePath: item.FilePathWithoutExtension);
-                PageSelectorModel.PopulateDefaults(routeModel, routeModel.ViewEnginePath, routeTemplate);
+                var isFallbackRoute = FallbackRazorPagesConventions.TryGetSupersedingPath(item.FilePathWithoutExtension, out var supersedingPath);
+                var viewEnginePath = isFallbackRoute ? supersedingPath : item.FilePathWithoutExtension;
+                var routeModel = new PageRouteModel(relativePath, viewEnginePath);
+                PageSelectorModel.PopulateDefaults(routeModel, viewEnginePath, routeTemplate);
+
+                if (isFallbackRoute)
+                {
+                    routeModel.Properties[FallbackRazorPagesConventions.FallbackPropertyKey] = true;
+                }
+
```

Comment:
There's a lot more functionality here and it keeps growing. Should we make a new OMNI-do-the-thing piece of code?